pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -330,18 +357,77 @@ stdenv.mkDerivation ({
     ''}
     ${optionalString doCoverage "mkdir -p $out/share && cp -r dist/hpc $out/share"}
     ${optionalString (enableSharedExecutables && isExecutable && !isGhcjs && stdenv.isDarwin && stdenv.lib.versionOlder ghc.version "7.10") ''
-      for exe in "$out/bin/"* ; do
-        install_name_tool -add_rpath "$out/lib/ghc-${ghc.version}/${pname}-${version}" "$exe"
+      for exe in "${binDir}/"* ; do
+        install_name_tool -add_rpath "${libDir}/${pname}-${version}" "$exe"
       done
     ''}
 
     ${optionalString enableSeparateDocOutput ''
-    for x in ${docdir "$doc"}/html/src/*.html; do
-      remove-references-to -t $out $x
+    # Remove references back to $out but also back to $lib if we have
+    # docs. $lib is needed as it stores path to haddock interfaces in the
+    # conf file which creates a cycle if docs refer back to library
+    # path.
+    mkdir -p ${docDir}
+
+    for x in ${docDir}/html/src/*.html; do
+      remove-references-to -t $out -t ${libDir} -t ${binDir} ${optionalString enableSeparateDataOutput "-t $data"} $x
     done
-    mkdir -p $doc
     ''}
-    ${optionalString enableSeparateDataOutput "mkdir -p $data"}
+
+    ${optionalString hasLibOutput ''
+    # Even if we don't have binary output for the package, things like
+    # Paths files will embed paths to bin/libexec directories in themselves
+    # which results in .lib <-> $out cyclic store reference. We
+    # therefore patch out the paths from separate library if we don't have
+    # separate bin output too.
+    #
+    # If we _do_ have separate bin and lib outputs, we may still be in
+    # trouble in case of shared executables: executable contains path to
+    # .lib, .lib contains path (through Paths) to .bin and we have a
+    # cycle.
+    #
+    # Lastly we have to deal with references from .lib back into
+    # $out/share if we're not splitting out data directory.
+    #
+    # It may happen that we have hasLibOutput set but the library
+    # directory was not created: this happens in the case that library
+    # section is not exposing any modules. See "fail" package for an
+    # example where no modules are exposed for GHC >= 8.0.
+    if [ -d ${libDir} ]; then
+      find ${libDir} -type f -exec \
+        remove-references-to -t ${binDir} -t ${libexecDir} "{}" \;
+    fi
+    ''}
+
+    ${optionalString (hasLibOutput && ! enableSeparateDocOutput) ''
+    # If we don't have separate docs, we have to patch out the ref to
+    # docs in package conf. This will likely break Haddock
+    # cross-package links but is necessary to break store cycle…
+    find ${libDir}/ -type f -name '*.conf' -exec \
+      remove-references-to -t ${docDir} "{}" \;
+    ''}
+
+    ${optionalString (hasLibOutput && ! enableSeparateDataOutput) ''
+    # Just like for doc output path in $out potentially landing in
+    # *.conf, we have to also remove the data directory so that it
+    # doesn't appear under data-dir field creating a cycle.
```

Comment:
this looks to me like it has potential for silently breaking packages: if a library needs access to data files and `enableSeparateDataOutput` is not set, we will brick the library, right? (because references will be replaced with non-existent files). 

Can we limit this to only `.conf` files, and have the build break in other cases?

Reply:
Good catch.